libyul/AST.h

```diff
@@ -41,7 +43,41 @@ using TypedNameList = std::vector<TypedName>;
 
 /// Literal number or string (up to 32 bytes)
 enum class LiteralKind { Number, Boolean, String };
-struct Literal { langutil::DebugData::ConstPtr debugData; LiteralKind kind; YulString value; Type type; };
+/// Literal value that holds a u256 word of data, can be of LiteralKind type and - in case of arguments to
+/// builtins - exceed the u256 word (32 characters), in which case the value is set to `unlimited`
+class LiteralValue {
+public:
+	using Data = u256;
+	using RepresentationHint = std::shared_ptr<std::string>;
+
+	LiteralValue() = default;
+	LiteralValue(Data const& _data, std::optional<std::string> const& _hint = std::nullopt, bool _unlimited = false):
+		m_data(_data), m_hint(_hint ? std::make_shared<std::string>(*_hint) : nullptr), m_unlimited(_unlimited) {}
+
+	bool operator==(LiteralValue const& _rhs) const
+	{
+		if (m_unlimited && _rhs.m_unlimited)
+		{
+			return *m_hint == *_rhs.m_hint;
+		}
+		return m_unlimited == _rhs.m_unlimited && m_data == _rhs.m_data;
+	}
+	bool operator==(Data const& _rhs) const { return m_data == _rhs; }
+	bool operator!=(Data const& _rhs) const { return m_data != _rhs; }
+
+	Data const& value() const { return m_data; }
+	bool unlimited() const { return m_unlimited; }
+
+	Data const& operator()() const { return value(); }
+
+	RepresentationHint const& hint() const { return m_hint; }
+
+private:
+	Data m_data {0};
+	RepresentationHint m_hint {nullptr};
+	bool m_unlimited {};
```

Comment:
These names at first seem straightforward: `m_data` is surely the content and `m_hint` is just an optional formatting thing, right? But then you realize they aren't. Sometimes `m_data` is irrelevant and it's the hint that stores the actual value. Or you can have weird states like unlimited but no hint (what does that even mean?). You can also have unlimited values with irrelevant `m_data` which does not match the string.

I think they at least need names that better reflect what they really are. But actually, maybe it would be even better to also change how it's stored so that fewer weird states are possible:
```c++
optional<u256> m_numericValue;
std::shared_ptr<std::string> m_stringValue;
```
The requirement (enforced with asserts) would be that at least one of these must be provided. And if they're both provided, their values must be consistent. Not having a numeric value means the string is no longer bound by the consistency rule and can be of arbitrary length.

What do you think?

Reply:
BTW, our usual style is `m_data{0}` (or, better `m_data = 0`). But the `bool` is the only POD type here so only it needs explicit initialization.

```suggestion
	Data m_data;
	RepresentationHint m_hint;
	bool m_unlimited{};
```

Reply:
I partially agree :) it's kind of dual purpose. you have the following cases:
- just the numerical value as u256 word
- numerical value + string value / representation hint, in that case the hint is reproduced upon output
- unlimited literal with just the representation hint == being its value

so i think the current structure fits the first two points nicely (as in: data + optional hint) but the third point is a bit of a special case which reuses the existing structure but doesn't really fit semantically. Admittedly it took a while for me to understand that the third point was even a thing. :D
To do this more cleanly I would actually prefer something like

```cpp
variant<string, tuple<u256, shared_ptr<string>>>
```

where the first variant is for literal arguments of builtins and the last for the other cases.

Reply:
Sure, a variant like that sounds good too.

Reply:
The variant gave me weird false positive maybe-uninitialized warnings treated as errors so I opted for your original suggestion after all.